lib/logflare/alerting.ex

```diff
@@ -0,0 +1,242 @@
+    user
+    |> Ecto.build_assoc(:alert_queries)
+    |> Repo.preload(:user)
+    |> AlertQuery.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Updates a alert_query.
+
+  ## Examples
+
+      iex> update_alert_query(alert_query, %{field: new_value})
+      {:ok, %AlertQuery{}}
+
+      iex> update_alert_query(alert_query, %{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def update_alert_query(%AlertQuery{} = alert_query, attrs) do
+    alert_query
+    |> Repo.preload(:user)
+    |> AlertQuery.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @doc """
+  Deletes a alert_query.
+
+  ## Examples
+
+      iex> delete_alert_query(alert_query)
+      {:ok, %AlertQuery{}}
+
+      iex> delete_alert_query(alert_query)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def delete_alert_query(%AlertQuery{} = alert_query) do
+    Repo.delete(alert_query)
+  end
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking alert_query changes.
+
+  ## Examples
+
+      iex> change_alert_query(alert_query)
+      %Ecto.Changeset{data: %AlertQuery{}}
+
+  """
+  @spec change_alert_query(AlertQuery.t()) :: Ecto.Changeset.t()
+  def change_alert_query(%AlertQuery{} = alert_query, attrs \\ %{}) do
+    AlertQuery.changeset(alert_query, attrs)
+  end
+
+  @doc """
+  Retrieves a Citrine.Job based on AlertQuery.
+  Citrine.Job shares the same id as AlertQuery, resulting in a 1-1 relationship.
+  """
+  @spec get_alert_job(AlertQuery.t()) :: Citrine.Job.t()
+  def get_alert_job(%AlertQuery{id: id}), do: get_alert_job(id)
+
+  def get_alert_job(id) do
+    case Logflare.AlertsScheduler.get_job(id) do
+      {_pid, job} -> job
+      nil -> nil
+    end
+  end
+
+  @doc """
+  Updates or creates a new Citrine.Job based on a given AlertQuery
+  """
+  @spec upsert_alert_job(AlertQuery.t()) :: {:ok, Citrine.Job.t()}
+  def upsert_alert_job(%AlertQuery{} = alert_query) do
+    Logflare.AlertsScheduler.put_job(%Citrine.Job{
+      id: alert_query.id,
+      schedule: alert_query.cron,
+      extended_syntax: false,
+      task: {:run_alert, [alert_query]}
+    })
+
+    {:ok, get_alert_job(alert_query)}
+  end
+
+  @doc """
+  Initializes and ensures that all alert jobs are created.
+  TODO: batching instead of loading whole table.
+  """
+  def init_alert_jobs do
+    AlertQuery
+    |> Repo.all()
+    |> Stream.each(fn alert_query ->
+      if get_alert_job(alert_query) == nil do
+        upsert_alert_job(alert_query)
+      end
+    end)
+    |> Stream.run()
+
+    :ok
+  end
+
+  @doc """
+  Performs the check lifecycle of an AlertQuery.
+
+  Send notifications if necessary configurations are set. If no results are returned from the query execution, no alert is sent.
+  """
+  @spec run_alert(AlertQuery.t()) :: :ok
+  def run_alert(%AlertQuery{} = alert_query) do
+    alert_query = alert_query |> Repo.preload([:user])
+
+    with {:ok, [_ | _] = results} <- execute_alert_query(alert_query) do
```

Comment:
Ah nice so if it returns >0 results it'll alert with the results.

Reply:
So what would my query be for say "send me an alert if my percentage of error logs was > 5% in the last 5 minutes?"

Reply:
```sql
select
t.request_url,
 countif(t.error = 'error') as errors
`my.logs` t
where timestamp_sub(current_timestamp(), INTERVAL 5 MINUTE)
group by t.request_url
having ( countif(t.error = 'error') / count(t.id) ) > 0.05
```
Something like this